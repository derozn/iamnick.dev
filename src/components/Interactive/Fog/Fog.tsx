import React, { useState, useRef, useEffect } from 'react';
import { useLoader, useFrame, extend } from 'react-three-fiber';
import { TextureLoader } from 'three';

import useStore from '#store';

import { CloudsMaterial } from './materials/clouds';
import { Clouds } from './objects/clouds';

extend({ CloudsMaterial, Clouds });

const Fog = () => {
  const clouds = useRef<Clouds>(null);
  const [time, setTime] = useState(0.0);
  const { textureUrl } = useStore((state) => state.fog);
  const cloudTexture = useLoader(TextureLoader, textureUrl);

  useEffect(() => {
    if (clouds.current) clouds.current.create();
  }, []);

  useFrame((_, delay) => {
    setTime(time + delay);
  });

  return (
    <mesh name="fog">
      <clouds args={[20]} ref={clouds} attach="geometry" />
      <cloudsMaterial
        attach="material"
        cloudTexture={cloudTexture}
        time={time}
        extensions-derivatives
      />
    </mesh>
  );
};

export default Fog;
