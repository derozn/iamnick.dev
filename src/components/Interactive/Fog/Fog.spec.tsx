import React, { Suspense } from 'react';
import { create } from 'react-test-renderer';
import { useLoader } from 'react-three-fiber';
import { TextureLoader } from 'three';

import useStore from '#store';
import Fog from './Fog';

jest.mock('#store');
jest.mock('react-three-fiber');

const mockUseLoader = useLoader as jest.Mock<any>;
const mockUseStore = useStore as jest.Mock<any>;

describe('components/Fog', () => {
  beforeEach(() => {
    jest.clearAllMocks();
  });

  it('renders Fog', () => {
    mockUseStore.mockImplementation((cb) => {
      return cb({ fog: { textureUrl: 'texture-url' } });
    });

    mockUseLoader.mockImplementationOnce(() => 'texture');

    const wrapper = create(
      <Suspense fallback={null}>
        <Fog />
      </Suspense>,
    );

    expect(mockUseLoader).toHaveBeenCalledWith(TextureLoader, 'texture-url');
    expect(wrapper.toJSON()).toMatchSnapshot();
  });
});
