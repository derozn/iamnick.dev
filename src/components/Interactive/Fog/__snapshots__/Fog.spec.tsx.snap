// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`components/Fog renders Fog 1`] = `
<mesh
  name="fog"
>
  <instancedBufferGeometry
    attach="geometry"
  />
  <fogMaterial
    attach="material"
    extensions-derivatives={true}
    fogTexture="texture"
    time={0}
  />
</mesh>
`;
